stdlib/public/core/AutoDiff.swift

```diff
@@ -495,6 +495,251 @@ public func gradient<T, U, V, R>(
   return { x, y, z in gradient(at: x, y, z, in: f) }
 }
 
+//===----------------------------------------------------------------------===//
+// Type-erased `AnyDerivative`
+//===----------------------------------------------------------------------===//
+
+internal protocol _AnyDerivativeBox {
+  // `Equatable` requirements (implied by `AdditiveArithmetic`).
+  func _isEqual(to other: _AnyDerivativeBox) -> Bool
+  func _isNotEqual(to other: _AnyDerivativeBox) -> Bool
+
+  // `AdditiveArithmetic` requirements.
+  static var _zero: _AnyDerivativeBox { get }
+  static var _cotangentZero: _AnyDerivativeBox { get }
+  func _adding(_ x: _AnyDerivativeBox) -> _AnyDerivativeBox
+  func _subtracting(_ x: _AnyDerivativeBox) -> _AnyDerivativeBox
+
+  // `Differentiable` requirements.
+  var _allDifferentiableVariables: _AnyDerivativeBox { get }
+  func _moved(along direction: _AnyDerivativeBox) -> _AnyDerivativeBox
+  func _tangentVector(from cotangent: _AnyDerivativeBox) -> _AnyDerivativeBox
+
+  /// The underlying base value, type-erased to `Any`.
+  var _typeErasedBase: Any { get }
+
+  /// Returns the underlying value unboxed to the given type, if possible.
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+}
+
+extension _AnyDerivativeBox {
+  /// Returns true if the underlying value has type `AnyDerivative.OpaqueZero`.
+  func _isOpaqueZero() -> Bool {
+    return _unboxed(to: AnyDerivative.OpaqueZero.self) != nil
+  }
+}
+
+@inline(never)
+@usableFromInline
+internal func _derivativeTypeMismatch(
+  _ x: Any, _ y: Any, file: StaticString = #file, line: UInt = #line
```

Comment:
It would be slightly more efficient to make this function take `Any.Type` instead of `Any`. Prevents unnecessary boxing of values when you just need their types.
```suggestion
  _ x: Any.Type, _ y: Any.Type, file: StaticString = #file, line: UInt = #line
```

Reply:
Done.